guide/src/async-await.md

```diff
@@ -0,0 +1,240 @@
+# Async / Await
+
+Both Python and Rust have support for async functions, but bindings for these functions are not as 
+straightforward as they are for blocking functions. Both languages have a fairly distinct model for 
+async functions and Python's needs in particular can sometimes be restrictive.
+
+[pyo3-asyncio](https://github.com/awestlake87/pyo3-asyncio) was created to provide conversions 
+between async Python and async Rust as well as manage the nitty gritty details of Python's event 
+loop.
+
+## Awaiting an Async Python Function in Rust
+
+Let's take a look at a dead simple async Python function:
+
+```python
+async def py_sleep():
+    await asyncio.sleep(1)
+```
+
+This function simply sleeps for 1 second and returns. So what does this look like to PyO3? 
+
+First, it helps to have a little background on Python's async functions. 
+Async functions in python are simply functions that return a `coroutine` object. You can read more about 
+`coroutine` objects in the [Python 3 docs](https://docs.python.org/3/library/asyncio-task.html), but
+for our purposes, we really don't need to know much about them. The key factor here is that calling
+an `async` function is _just like calling a regular function_, the only difference is that we have
+to do something special with the object that it returns.
+
+Normally in Python, that something special is the `await` keyword, but in Rust, we don't have the 
+luxury of using Python's syntax. Luckily, Rust also has an `await` keyword that does something
+similar, we just need to find a way of converting a `&PyAny` into a Rust future so we can use the
+`await` keyword on it.
+
+That's where pyo3-asyncio comes in. `pyo3_asyncio::into_future` performs this conversion for us:
+
+```rust
+let future = Python::with_gil(|py| {
+    // import the module containing the py_sleep function
+    let example = py.import("example")?;
+
+    // calling the py_sleep method like a normal function returns a coroutine
+    let coroutine = example.call_method0("py_sleep")?;
+
+    // convert the coroutine into a Rust future
+    pyo3_asyncio::into_future(py, coroutine)
+})?;
+
+// await the future
+future.await;
+```
+
+## Awaiting a Rust Future in Python
+
+Here we have the same async function as before written in Rust using the 
+[`async-std`](https://async.rs/) runtime:
+
+```rust
+async fn rust_sleep() {
+    async_std::task::sleep(Duration::from_secs(1)).await;
+}
+```
+
+Similar to Python, Rust's async functions also return a special object called a
+`Future`:
+
+```rust
+let future = rust_sleep();
+```
+
+We can convert this future object into Python to make it "Awaitable". This tells Python that you can
+use the `await` keyword with it. In order to do this, we'll call 
+`pyo3_asyncio::async_std::into_coroutine`:
+
+```rust
+#[pyfunction]
+fn call_rust_sleep(py: Python) -> PyResult<PyObject> {
```

Comment:
The return type here would actually be `asyncio.Future` here since the event loop waits on us to run the future on the Rust event loop and call `asyncio.Future.set_result`.

`into_coroutine` is a bit of a misnomer, but I wasn't really sure what to call it honestly. `into_future` is already taken, maybe `into_awaitable` would be a better name?